File: code/java/tree/src/main/java/org/swufe/KDTree.java
package org.swufe;

import java.util.*;

class Point {
    private int x;
    private int y;

    public Point(int x, int y) {
        this.x = x;
        this.y = y;
    }

    public int getX() {
        return x;
    }

    public int getY() {
        return y;
    }

    public Integer getItem(int i) {
        assert i == 0 || i == 1;
        if (i == 0) return x;
        else return y;
    }

    @Override
    public String toString() {
        return "Point{" +
                "x=" + x +
                ", y=" + y +
                '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Point point = (Point) o;
        return Double.compare(point.x, x) == 0 && Double.compare(point.y, y) == 0;
    }

    @Override
    public int hashCode() {
        return Objects.hash(x, y);
    }
}

class Rectangle {
    private Point lower;
    private Point upper;

    public Rectangle(Point lower, Point upper) {
        this.lower = lower;
        this.upper = upper;
    }

    public Point getLower() {
        return lower;
    }

    public Point getUpper() {
        return upper;
    }

    public boolean isContain(Point p) {
        return this.lower.getX() <= p.getX() && this.upper.getX() >= p.getX() &&
                this.lower.getY() <= p.getY() && this.upper.getY() >= p.getY();
    }

    @Override
    public String toString() {
        return "Rectangle{" +
                "lower=" + lower +
                ", upper=" + upper +
                '}';
    }
}

public class KDTree {

    private static class Node {
        private Point location;
        private Node left;
        private Node right;

        public Node(Point location, Node left, Node right) {
            this.location = location;
            this.left = left;
            this.right = right;
        }
    }

    private Node root;
    private int size;

    public KDTree() {
        root = null;
        size = 0;
    }

    public void insert(List<Point> p) {
        throw new UnsupportedOperationException("Not supported yet.");
    }

    public List<Point> range(Rectangle rectangle) {
        throw new UnsupportedOperationException("Not supported yet.");
    }

    public static void main(String[] args) {
        List<Point> points = new ArrayList<>();
        for (int i = 0; i < 2000; i++) {
            for (int j = 0; j < 2000; j++) {
                points.add(new Point(i, j));
            }
        }

        Point lower = new Point(500, 500);
        Point upper = new Point(504, 504);
        Rectangle rectangle = new Rectangle(lower, upper);
        // naive method
        double start = System.currentTimeMillis();
        List<Point> result1 = points.stream().filter(rectangle::isContain).toList();
        double end = System.currentTimeMillis();
        System.out.printf("Naive method: %fms\n", end - start);

        KDTree kdTree = new KDTree();
        kdTree.insert(points);
        // kd tree
        start = System.currentTimeMillis();
        List<Point> result2 = kdTree.range(rectangle);
        end = System.currentTimeMillis();
        System.out.printf("kd tree: %fms\n", end - start);
    }
}

File: code/python/tree/kd_tree.py
from typing import List
from collections import namedtuple
import time


class Point(namedtuple("Point", "x y")):
    def __repr__(self) -> str:
        return f'Point{tuple(self)!r}'


class Rectangle(namedtuple("Rectangle", "lower upper")):
    def __repr__(self) -> str:
        return f'Rectangle{tuple(self)!r}'

    def is_contains(self, p: Point) -> bool:
        return self.lower.x <= p.x <= self.upper.x and self.lower.y <= p.y <= self.upper.y


class Node(namedtuple("Node", "location left right")):
    """
    location: Point
    left: Node
    right: Node
    """

    def __repr__(self):
        return f'{tuple(self)!r}'


class KDTree:
    """k-d tree"""

    def __init__(self):
        self._root = None
        self._n = 0

    def insert(self, p: List[Point]):
        """insert a list of points"""
        pass

    def range(self, rectangle: Rectangle) -> List[Point]:
        """range query"""
        pass


def range_test():
    points = [Point(7, 2), Point(5, 4), Point(9, 6), Point(4, 7), Point(8, 1), Point(2, 3)]
    kd = KDTree()
    kd.insert(points)
    result = kd.range(Rectangle(Point(0, 0), Point(6, 6)))
    assert sorted(result) == sorted([Point(2, 3), Point(5, 4)])


def performance_test():
    points = [Point(x, y) for x in range(1000) for y in range(1000)]

    lower = Point(500, 500)
    upper = Point(504, 504)
    rectangle = Rectangle(lower, upper)
    #  naive method
    start = int(round(time.time() * 1000))
    result1 = [p for p in points if rectangle.is_contains(p)]
    end = int(round(time.time() * 1000))
    print(f'Naive method: {end - start}ms')

    kd = KDTree()
    kd.insert(points)
    # k-d tree
    start = int(round(time.time() * 1000))
    result2 = kd.range(rectangle)
    end = int(round(time.time() * 1000))
    print(f'K-D tree: {end - start}ms')

    assert sorted(result1) == sorted(result2)


if __name__ == '__main__':
    range_test()
    performance_test()
